### ISSUE TITLE
Let customers install apps on the management cluster

### ISSUE THREAD

**Issue by teemow** [giantswarm/roadmap#1073]
## User Story

 - As a customer I want to install managed apps on the management cluster so that I can add central functionality that is available for all workload clusters

Examples:
 * Harbor
 * Crossplane
 * ArgoCD

## Details, Background

Some automation/management functionality might be optional from our perspective. It would be good to let customers install (managed) apps to avoid that customers have to create second class management clusters (eg use a workload cluster as an "automation" cluster) that adds a lot of complexity in terms of security and networking (needs peering to all other workload clusters). 

This would also allow us to let customers play with tools already that are on our roadmap but not yet ready to become fully managed by us on the management cluster.

We need to ensure that there is isolation from sensitive functionality and secrets. But enabling the platform teams at the customer is more important than preventing any problem in the first place. Customers can also destroy workload clusters but we've decided to value freedom and flexibility more.

**Comment by gianfranco-l** [comment]
@giantswarm/team-honeybadger We are doing careful (baby) steps towards this big scope with discovery work for a specific customer case. cc @MarcelMue can add more details if needed. We will keep an eye on this as a larger story while we continue doing discovery work for this specific customer and future customers that might be interested in this as well

**Comment by JosephSalisbury** [comment]
We chatted in SIG Product WH, @cornelius-keller brought up that this is getting important for THG, they want to run their own operators to create OpenStack infrastructure.

There's a discussion about using Crossplane, and whether it's okay if they're running their own operators.

Let's chat about this one further next week and get to a decision on this one.

**Comment by cornelius-keller** [comment]
THG wants to provision infrastructure together with a cluster that is being created. They want to have have the CRs in the same commit and repo as the cluster definition. They are currently developing operators to do this, currently without crossplane.

**Comment by teemow** [comment]
I am sure we will see more use cases in the future where it makes sense to extend the management api. And I think in the first place we should not stand in the way of our customers and allow them to install apps. Otherwise we'll always be the bottleneck. 

There are many examples. Customers want to install Harbor, Crossplane, ArgoCD or Backstage on the management cluster. Let's show them how to do this and enable them. Once we have capacity and time we can take over helm chart management and operations. But let's not block the customers in the first place.

And yes, we need to make sure this is secure and robust.

**Comment by MarcelMue** [comment]
I think by far the biggest concern we currently have is _how_ to do it securely and robust. It feels to me like RBAC will not be enough to make it work because most of the apps which were mentioned by Timo expect some kind of elevated privilege to run. For example ArgoCD will not work at all without giving it very high privileges. Crossplane needs cluster wide access to all the CRs it manages.

So what could even be a realistic approach to make it possible for customers to install these apps? I feel currently that this would not work in a multi tenant MC.

**Comment by teemow** [comment]
I think we are fine disabling this for multi-tenant MCs.

What we need to find out then is:

1. How to create a user role that is allowed to install apps on MCs and how to create roles that are not allowed to do this and don't need such high privileges?
2. In case of high privileges which (giant swarm) secrets can be read and how can we make sure that the scope of these secrets is exactly one customer? This way we don't need to worry about higher privileges on MCs imo.

Does that make sense?

**Comment by MarcelMue** [comment]
I think 1. is the real issue though:
No matter how we create the roles, some of these apps simply do not work with lowered permissions. We could try to bring changes to the respective upstream projects but the effort would likely be huge. 

To me 2. seems to be the more sensible approach then, it's not as nice but it seems more realistic to me in a reasonable amount of effort / timeframe. Would love to hear other @giantswarm/team-honeybadger opinions though.

**Comment by teemow** [comment]
@MarcelMue ok let's make sure to not mix up gitops and MAPI. 

With 1. I was talking about MAPI. We just need to make sure that people can access the management cluster to install an app on a WC without being able to install something on the MC and without risk that secrets leak. 

The gitops pipeline account/role is a completly different beast. This is where 2. comes in.

**Comment by MarcelMue** [comment]
You have lost me:
```
1. How to create a user role that is allowed to install apps on MCs and how to create roles that are not allowed to do this and don't need such high privileges?
```
This to me means installing an app on the MC where the pods of the app are running on the MC.
```
With 1. I was talking about MAPI. We just need to make sure that people can access the management cluster to install an app on a WC without being able to install something on the MC and without risk that secrets leak.
```
This is something we already have IMO. People can install their own apps on their WC using MAPI. 

I didn't relate anything on GitOps: just simply if we have an app created by a customer on the MC where the actual application is supposed to be running on the MC then we have large permission issues (because most upstream apps/controllers in general ever assume they could be running in a restricted permission environment).

**Comment by teemow** [comment]
I was not sure if we can distinguish between apps on MC and apps on WC already with RBAC.

**Comment by puja108** [comment]
We talked about this today with Honeybadger, and decided we need to at least discuss this again in SIG Product as we are seeing more and more customer requests for running central management tools that we currently don't offer (yet). It is already scheduled for a future SWH, but as those recently got skipped a lot and vacation time is upcoming, I'd like to at least quickly discuss in the next sync and also kick-off async thoughts here.

We know the impact on not only Honeybadger, but also on KaaS teams that are responsible for the MCs.

We also know that there's some blockers in terms of security and MC setup as well as maybe thoughts towards better isolation.

Discussing and thinking about this feature soon is critical so we keep its implications in mind while working on the new MC setup for CAPI (cc @cornelius-keller @alex-dabija), e.g. in the way we create and manage our secrets.

**Comment by puja108** [comment]
Thinking about this a bit more we might need to also scope the different parts of this epic. Something like:
1. MCs that enable customer deployments through a secure and scalable setup (KaaS)
2. A safe and reliable way for customers to make such deployments (Honeybadger DX)

Having the parts clear makes it easier to think about the respective work and effort for each team.

**Comment by puja108** [comment]
Fine to close in favor of something specific which would represent the future coverage of the feature.

**Comment by teemow** [comment]
Do we have docs how to do this? We have other customers that are interested in the feature too.

**Comment by QuentinBisson** [comment]
Monitoring of those apps is also to be taken into account here and so far has not been discussed

**Comment by kubasobon** [comment]
Yes, of course. The existing rules will also have to be updated (DeploymentNotSatisfied for example).

**Comment by QuentinBisson** [comment]
I am mentioning that because one customer wants to be able to create their own service monitor in the MC and maybe deploy a prometheus and alertmanager there and this will collide with our own setup as we currently do not restrict where to look for rules and service monitors as it would be counter productive

**Comment by teemow** [comment]
Let's also discuss having those metrics in our MC prom and creating a section in Grafana where customers can create their own dashboards. Imo this would be much more convenient. Bonus points if the customer can also add alertmanager routing and prometheus rules for their own metrics :smile: .

**Comment by teemow** [comment]
@QuentinBisson can you discuss this in Atlas? Maybe it makes sense to split off an issue. Happy to help with the specs or to setup a meeting with the customer.

**Comment by QuentinBisson** [comment]
Definitely, I created [this issue](https://github.com/giantswarm/giantswarm/issues/23411) that we need to refine and prioritize. Did we announce an ETA on this feature to customers? Once we refine it, let's setup a meeting with the customer :)

**Comment by teemow** [comment]
No eta yet. They might be fine for now with the workaround you proposed using remote write. We haven't done any research how many customers would want this too.

Letting customers add their own dashboards apart from apps on the MC is definitely something to look at separately too. This might be interesting for special cases/clusters where our dashboards aren't enough.

**Comment by QuentinBisson** [comment]
You are right and it will definitely be if we monitor their apps and it's actually possible today but we probably forgot to communicate that at least internally. Documenting it in the intranet first would be a good thing

**Comment by teemow** [comment]
@weatherhog @piontec what is left todo here?

**Comment by piontec** [comment]
> @weatherhog @piontec what is left todo here?

It depends.

If you mean "to install some apps, that don't need clsuter-admin permissions and that we probably have to help with", then nothing - we've done that already.

If you mean general functionality, where customers are finally cluster-admins (without major security problems), it's tracked by [this ticket](https://github.com/giantswarm/roadmap/issues/2457), which in turn is blocked by [this one](https://github.com/giantswarm/roadmap/issues/2456). As this solves a lot of headaches for us as well, we've already started working on it.

**Comment by teemow** [comment]
Alright. Thanks. I've added the issues to a tasklist in the description.

**Comment by weatherhog** [comment]
@piontec what is missing to proceed here as it seems all the headaches are solved.

**Comment by piontec** [comment]
No, they're not. Installing any app effectively means cluster admin priviledges, otherwise we have to manually tune all the RBAC rules to make it work. So, we want to give customers cluster admin to make this easier, but that is blocked by the security risk of cross-installations shared secrets.

**Comment by QuentinBisson** [comment]
Do we need cluster admin to create HelmRelease CRs? Is there a reason we cannot create a rôle that only does that ?

**Comment by puja108** [comment]
It's the same risk if the HelmRelease can contain anything and extract secrets from the cluster.

I'd love to remove the requirement of needing full cluster-admin here. What are the use-cases of not limiting customers to certain namespaces? Is it just for cluster-level CRDs? What were the use cases we encountered that needed that?

**Comment by puja108** [comment]
One idea we had once was to give customers access to a GitOps repo for their own MC and controlling deployments to the MC through review by us. For the shared ownership we have, I'd still like to limit the blast-radius of customer access. At least it would be a good first step.

The requirement for removing shared secrets is still a good one, but we could handle it separately (also wrt our goals of maybe at some point being able to have customers fully manage their own GS)

**Comment by piontec** [comment]
Cluster admin permissions are not needed to create a HR: that's exactly how things work right now. Customers can create HRs on MC (I know, obvious) and they can actually target the local - MC - cluster, but we enforce `impersonation account` in each HR to be the `automation` ServiceAccount. But this means "the game is on": we have to think all the time how to restrict this account and how it can be used for exploiting the access. And we can't enable any CRD management and... the list of problems goes on.

I like the idea of reviewed configs for github PRs, but:
- this can put potentialy a lot of load on teams, when they effectively have to do security audit of each of such PRs,
- it's easy to miss something anyway

On the other hand, I'm pretty sure that if you dig long enough, you can get cluster admin right now anyway.

My suggestion: let's do soft blocking - block customers from accidentally destroying ie. stuff in the `giantswarm` namespace, but don't try to block every possible exploitation. In other words: protect from accidents, but not from deliberate exploitation.
